packages/babel-types/src/clone/cloneNode.ts

```diff
@@ -85,6 +85,34 @@ function cloneNodeInternal<T extends t.Node>(
   } else if (!hasOwn(NODE_FIELDS, type)) {
     throw new Error(`Unknown node type: "${type}"`);
   } else {
+    if (type === "Placeholder") {
+      // @ts-expect-error undocumented
+      if (hasOwn(node, "optional") && typeof node.optional === "boolean") {
+        // @ts-expect-error undocumented
+        newNode.optional = node.optional;
+      }
+
+      if (hasOwn(node, "typeAnnotation")) {
+        newNode.typeAnnotation = deep
```

Comment:
Can we declare these various extra properties in the type definitions of Placehodler?

Reply:
```
export const PLACEHOLDERS = [
  "Identifier",
  "StringLiteral",
  "Expression",
  "Statement",
  "Declaration",
  "BlockStatement",
  "ClassBody",
  "Pattern",
] as const;
```
I'm afraid there are other properties that should be here, which would result in a large number of properties in the `Placehodler` type definition.
However, I haven't thought of a better way so far.

Reply:
Can they, though? For example, a placeholder will never have a `.body` because it's "inside" the node.

The identifier properties are special because they are not really _inside_ the identifier, so they are not hidden by the placeholder.